Enable adam foreach in full train step tracing

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97897
* #97853

Main changes:

1. Registered several foreach ops to both meta and DTensor
2. Skip redundant getitem node when expanding foreach ops with DTensor

mrshenli committed:
[WIP][Don't Review Yet] Enable adam foreach in full train step tracing

[ghstack-poisoned]

bdhirsh commented on torch/_meta_registrations.py:
```diff
@@ -1206,14 +1232,75 @@ def _check_foreach_binop_tensor_lists(self, other):
+
+@register_meta([aten._foreach_pow.ScalarAndTensor])
+def meta__foreach_pow_scalar_and_tensor(self, exponent):
+    check(isinstance(exponent, List), lambda: f"exponent must be a tensor list but got {type(exponent)}")
```
these all look ok to me.

bdhirsh commented on torch/_meta_registrations.py:
```diff
@@ -1206,14 +1232,75 @@ def _check_foreach_binop_tensor_lists(self, other):
+)
+def meta__foreach_addcop__scalar(self, tensor1, tensor2, scalar=1):
+    assert all([isinstance(l, List) for l in [self, tensor1, tensor2]]), (
+        "All arguments of _foreach_addcmul_ must be List[Tensor], but got "
```
Nit: don't hardcode `addcmul_` into the error string if you're re-using this meta kernel for multiple ops.

If we need the hardcoded string (to satisfy tests), you'll have to register the function separately.

mrshenli committed:
Update on "[WIP][Don't Review Yet] Enable adam foreach in full train step tracing"




[ghstack-poisoned]

mrshenli committed:
Update on "[WIP][Don't Review Yet] Enable adam foreach in full train step tracing"




[ghstack-poisoned]

mrshenli committed:
Update on "Enable adam foreach in full train step tracing"


Main changes:

1. Registered several foreach ops to both meta and DTensor
2. Skip redundant getitem node when expanding foreach ops with DTensor

[ghstack-poisoned]

mrshenli committed:
Update on "Enable adam foreach in full train step tracing"


Main changes:

1. Registered several foreach ops to both meta and DTensor
2. Skip redundant getitem node when expanding foreach ops with DTensor

[ghstack-poisoned]

wanchaol commented on torch/distributed/_spmd/experimental_ops.py:
```diff
@@ -17,17 +17,45 @@
         [isinstance(s, DTensorSpec) for s in self]
     )
+    # FIXME(@mrshenli): for sqrt, this is only mathematically correct for
+    # Replicate and Shard tensor.
```
oh do you mean it needs to guard on partial?

mrshenli replied:
not sure yet, we might need to return arg schema suggestions with either Replicate or Shard to make this sound, but it doesn't seem the optimal decision on Replicate vs Shard can be make locally here?

fegin reviewed (approved):
LGTM

mrshenli committed:
Update on "Enable adam foreach in full train step tracing"


Main changes:

1. Registered several foreach ops to both meta and DTensor
2. Skip redundant getitem node when expanding foreach ops with DTensor

[ghstack-poisoned]
